mozsvc/gunicorn_worker.py

```diff
@@ -0,0 +1,236 @@
+
+        * a timeout enforced on each individual request, rather than on
+          inactivity of the worker as a whole.
+
+        * a signal handler to dump memory usage data on SIGUSR2.
+
+    To detect eventloop blocking, the worker installs a greenlet trace
+    function that increments a counter on each context switch.  A background
+    (os-level) thread monitors this counter and prints a traceback if it has
+    not changed within a configurable number of seconds.
+    """
+
+    def init_process(self):
+        # Check if we need a background thread to monitor memory use.
+        needs_monitoring_thread = False
+        if MAX_MEMORY_USAGE:
+            self._last_memory_check_time = time.time()
+            needs_monitoring_thread = True
+
+        # Set up a greenlet tracing hook to monitor for event-loop blockage,
+        # but only if monitoring is both possible and required.
+        if hasattr(greenlet, "settrace") and MAX_BLOCKING_TIME > 0:
+            # Grab a reference to the gevent hub.
+            # It is needed in a background thread, but is only visible from
+            # the main thread, so we need to store an explicit reference to it.
+            self._active_hub = gevent.hub.get_hub()
+            # Set up a trace function to record each greenlet switch.
+            self._active_greenlet = None
+            self._greenlet_switch_counter = 0
+            greenlet.settrace(self._greenlet_switch_tracer)
+            self._main_thread_id = _real_get_ident()
+            needs_monitoring_thread = True
+
+        # Create a real thread to monitor out execution.
+        # Since this will be a long-running daemon thread, it's OK to
+        # fire-and-forget using the low-level start_new_thread function.
+        if needs_monitoring_thread:
+            _real_start_new_thread(self._process_monitoring_thread, ())
+
+        # Continue to superclass initialization logic.
+        # Note that this runs the main loop and never returns.
+        super(MozSvcGeventWorker, self).init_process()
+
+    def init_signals(self):
+        # Leave all signals defined by the superclass in place.
+        super(MozSvcGeventWorker, self).init_signals()
+
+        # Hook up SIGUSR2 to dump memory usage information.
+        # This will be useful for debugging memory leaks and the like.
+        signal.signal(signal.SIGUSR2, self._dump_memory_usage)
+        if hasattr(signal, "siginterrupt"):
+            signal.siginterrupt(signal.SIGUSR2, False)
+
+    def handle_request(self, *args):
+        # Apply the configured 'timeout' value to each individual request.
+        # Note that self.timeout is set to half the configured timeout by
+        # the arbiter, so we use the value directly from the config.
+        with gevent.Timeout(self.cfg.timeout):
+            return super(MozSvcGeventWorker, self).handle_request(*args)
+
+    def _greenlet_switch_tracer(self, what, xxx_todo_changeme):
```

Comment:
From the `startup.sh` script in this repo, it looks like shavar is using `uswgi` rather than `gunicorn` in production. If so then it may not be worth your time to chase down the necessary changes here, and instead just delete this worker and any tests for it from your local fork of `mozsvc`.